src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -15,27 +16,33 @@ public class CompilationWithAnalyzers
         private readonly Compilation _compilation;
         private readonly CancellationToken _cancellationToken;
         private readonly ConcurrentSet<Diagnostic> _exceptionDiagnostics;
+        private ImmutableHashSet<SyntaxTree> _analyzedSyntaxTrees = ImmutableHashSet<SyntaxTree>.Empty;
+        private Dictionary<SyntaxTree, Task<ImmutableArray<Diagnostic>>> _documentTasks = new Dictionary<SyntaxTree, Task<ImmutableArray<Diagnostic>>>();
+        private Task<ImmutableArray<Diagnostic>> _completeAnalysisTask;
+        private Task _latestAnalysisTask;
+        private readonly object _analysisLock = new object();
 
         public Compilation Compilation
         {
             get { return _compilation; }
         }
-
+        
         /// <summary>
         /// Creates a new compilation by attaching diagnostic analyzers to an existing compilation.
         /// </summary>
         /// <param name="compilation">The original compilation.</param>
         /// <param name="analyzers">The set of analyzers to include in future analyses.</param>
         /// <param name="options">Options that are passed to analyzers.</param>
+        /// <param name="onAnalyzerException">Action to invoke if an analyzer throws an exception.</param>
         /// <param name="cancellationToken">A cancellation token that can be used to abort analysis.</param>
-        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, CancellationToken cancellationToken)
+        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException, CancellationToken cancellationToken)
         {
             _cancellationToken = cancellationToken;
```

Comment:
In reality, we already have multiple analysis clients (error list, light bulb, ctrl + dot), each with a different priority. Anything explicitly invoked by the user, say ctrl + dot, must take precedence (shouldn't have to wait for any non-user triggered analysis to complete) and should be cancellable. Probably with the current design that means that each diagnostic client would need to have a separate CompilationWithAnalyzers instance with cloned compilation, which may or may not be acceptable based on performance analysis. However, I don't think disallowing cancellation is an option for quite a few of the diagnostic clients.